### PR TITLE
Properly reject non-json FromBody parameter binding

### LINKED ISSUE #35856: RequestDelegateFactory should return "415 Unsupported Media Type" response when parameter binding and the request isn't JSON
The implicit request body reading done by `RequestDelegateFactory` to populate `FromBody` parameters of the route handler blindly assumes the request body format is JSON, even when the request is sent with at `Content-Type` header that indicates the format is something other than JSON. This leads to a JSON deserialization exception being thrown if a request is sent with, e.g., XML, but the route handler relies on the default parameter binding logic.

Rather than always trying to deserialize the request body as JSON, the `Content-Type` header should first be checked and if it's not JSON, a "415 Unsupported Media Type" response should be returned. Note this only applies to the implicit/default request body deserialization logic, not in cases where the parameter is populated by a `BindAsync` method on the target type.

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/dotnet/aspnetcore/issues/35856

### CODE REVIEW

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 1]
```diff
@@ -1170,6 +1182,14 @@ public static void ParameterBindingFromHttpContextFailed(HttpContext httpContext
                 EventName = "ParameterBindingFromHttpContextFailed")]
             private static partial void ParameterBindingFromHttpContextFailed(ILogger logger, string parameterType, string parameterName);
 
+            public static void UnexpectedContentType(HttpContext httpContext, string? contentType)
+                => UnexpectedContentType(GetLogger(httpContext), contentType ?? "(none)");
+
+            [LoggerMessage(6, LogLevel.Debug,
+                @"Expected json content-type but got {ContentType}.",
```
Comment (DamianEdwards) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:1189]:
```suggestion
                @"Expected a supported JSON media type but got \"{ContentType}\".",
```

**Comment by BrennanConroy** [comment]
Did someone click "re-run failed jobs"?

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 2]
```diff
@@ -1170,6 +1182,14 @@ public static void ParameterBindingFromHttpContextFailed(HttpContext httpContext
                 EventName = "ParameterBindingFromHttpContextFailed")]
             private static partial void ParameterBindingFromHttpContextFailed(ILogger logger, string parameterType, string parameterName);
 
+            public static void UnexpectedContentType(HttpContext httpContext, string? contentType)
+                => UnexpectedContentType(GetLogger(httpContext), contentType ?? "(none)");
+
+            [LoggerMessage(6, LogLevel.Debug,
+                @"Expected a supported JSON media type but got \"{ContentType}\".",
```
Comment (BrennanConroy) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:1189]:
```suggestion
                "Expected a supported JSON media type but got \"{ContentType}\".",
```

**Review by davidfowl — APPROVED** [review]

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs [hunk 3]
```diff
@@ -2393,7 +2451,7 @@ private class CustomTodo : Todo
                 Assert.Equal(typeof(CustomTodo), parameter.ParameterType);
                 Assert.Equal("customTodo", parameter.Name);
 
-                var body = await context.Request.ReadFromJsonAsync<CustomTodo>();
+                var body = await JsonSerializer.DeserializeAsync<CustomTodo>(context.Request.Body);
```
Comment (davidfowl) [inline src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:2454]: ?

Comment (BrennanConroy) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:2454]: `ReadFromJsonAsync` throws for non json content type, and for testing I am sending Json without a json content type. I can make a separate type for the test to make it more focused.

**Pull request merged**